src/NumericalAlgorithms/Spectral/SwshCollocation.hpp

```diff
@@ -0,0 +1,256 @@
+// Helping functor to appropriately delete a stored `sharp_geom_info**`
+struct DestroySharpGeometry {
+  void operator()(sharp_geom_info** to_delete) noexcept {
+    sharp_destroy_geom_info(*to_delete);
+    free(to_delete);
+  }
+};
+}  // namespace detail
+/// \endcond
+
+/// A container for reporting a single collocation point for libsharp compatible
+/// data structures
+struct LibsharpCollocationPoint {
+  size_t offset;
+  double theta;
+  double phi;
+};
+
+/// \brief A wrapper class for the spherical harmonic library collocation data
+///
+/// \details The currently chosen library for spin-weighted spherical harmonic
+/// transforms is libsharp. The `Collocation` class stores the
+/// libsharp `sharp_geom_info` object, which contains data about
+/// 1. The angular collocation points used in spin-weighted spherical harmonic
+/// transforms
+/// 2. The memory representation of double-type values at those collocation
+/// points
+///
+/// \tparam Representation the ComplexRepresentation, either
+/// `ComplexRepresentation::Interleaved` or
+/// `ComplexRepresentation::RealsThenImags` compatible with the generated
+/// `Collocation` - this is necessary because the stored libsharp type contains
+/// memory stride information.
+template <ComplexRepresentation Representation>
+class Collocation {
+ public:
+  class CollocationIterator {
+   public:
+    /// create a new iterator. defaults to the start of the supplied object
+    explicit CollocationIterator(
+        gsl::not_null<Collocation<Representation>*> collocation,
+        size_t start_i = 0) noexcept
+        : index_{start_i}, collocation_{collocation} {}
+
+    /// recovers the data at the collocation point using a
+    /// `LibsharpCollocationPoint`, which stores the vector `offset` of the
+    /// location of the `theta`, `phi` point in libsharp compatible data
+    LibsharpCollocationPoint operator*() const noexcept {
+      return LibsharpCollocationPoint{index_, collocation_->theta(index_),
+                                      collocation_->phi(index_)};
+    }
+    /// advance the iterator by one position (prefix)
+    CollocationIterator& operator++() noexcept {
+      ++index_;
+      return *this;
+    };
+    /// advance the iterator by one position (postfix)
+    const CollocationIterator operator++(int) noexcept {
+      return CollocationIterator(collocation_, index_++);
+    }
+
+    /// retreat the iterator by one position (prefix)
+    CollocationIterator& operator--() noexcept {
+      --index_;
+      return *this;
+    };
+    /// retreat the iterator by one position (prefix)
+    const CollocationIterator operator--(int) noexcept {
+      return CollocationIterator(collocation_, index_--);
+    }
+
+    /// \cond HIDDEN_SYMBOLS
+    // prevent GCC warning about the non-template friend. This friend shouldn't
+    // be a template, it's a friend of the non-template iterator. Also, clang
+    // doesn't have that warning, so we need to avoid the warning about an
```

Comment:
but the parent class is a template, so I'm a bit worried about ignoring the warning.

You could just do a member `operator==`:
```cpp
bool operator==(const CollocationIterator& rhs) noexcept;
```